fdbrpc/DDSketch.h

```diff
@@ -0,0 +1,250 @@
+template <class T>
+class DDSketch {
+public:
+	explicit DDSketch(double errorGuarantee = 0.1)
+	  : errorGuarantee(errorGuarantee), gamma((1.0 + errorGuarantee) / (1.0 - errorGuarantee)), logGamma(log(gamma)),
+	    multiplier(FASTLOG ? fastLogger::correctingFactor * log(2) / log(1 + 2 * errorGuarantee / (1 - errorGuarantee))
+	                       : 1),
+	    offset((FASTLOG ? fastLogger::fastlog(1 / EPS) * multiplier : log(1 / EPS)) + 5), populationSize(0),
+	    zeroPopulationSize(0), minValue(T()), maxValue(T()), sum(T()) {
+		buckets.resize(2 * offset, 0);
+	}
+
+	DDSketch<T>& addSample(T sample) {
+		// Call it addSample for now, while it is not a sample anymore
+		if (!populationSize) minValue = maxValue = sample;
+
+		if (sample <= EPS) {
+			zeroPopulationSize++;
+		} else {
+#if FASTLOG
+			int index = ceil(fastLogger::fastlog(sample) * multiplier);
+#else
+			int index = ceil(log(sample) / logGamma);
+#endif
+			ASSERT(index + offset >= 0 && index + offset < buckets.size());
+			buckets[index + offset]++;
+		}
+
+		populationSize++;
+		sum += sample;
+		maxValue = std::max(maxValue, sample);
+		minValue = std::min(minValue, sample);
+		return *this;
+	}
+
+	double mean() const {
+		if (populationSize == 0) return 0;
+		return (double)sum / populationSize;
+	}
+
+	T median() { return percentile(0.5); }
+
+	T percentile(double percentile) { // Find the tPP-th (0-indexed) element
+		ASSERT(percentile >= 0 && percentile <= 1);
+
+		if (populationSize == 0) return T();
+		uint64_t targetPercentilePopulation = percentile * (populationSize - 1);
+		if (targetPercentilePopulation < zeroPopulationSize) return T(0);
+
+		int index = -1;
+		bool found = false;
+		if (percentile <= 0.5) { // count up
+			uint64_t count = zeroPopulationSize;
+			for (size_t i = 0; i < buckets.size(); i++) {
+				if (targetPercentilePopulation < count + buckets[i]) {
+					// count + buckets[i] = # of numbers so far (from the rightmost to this
+					// bucket, inclusive), so if target is in this bucket, it should means tPP < cnt + bck[i]
+					found = true;
+					index = i;
+					break;
+				}
+				count += buckets[i];
+			}
+		} else { // and count down
+			uint64_t count = 0;
+			for (size_t i = buckets.size(); i >= 0; i--) {
+				if (targetPercentilePopulation + count + buckets[i] >= populationSize) {
+					// cnt + bkt[i] is # of numbers to the right of this bucket (incl.),
+					// so if target is not in this bucket (i.e., to the left of this bucket),
+					// it would be as right as the left bucket's rightmost number, so
+					// we would have tPP + cnt + bkt[i] < total population (tPP is 0-indexed),
+					// that means target is in this bucket if this condition is not satisfied.
+					found = true;
+					index = i;
+					break;
+				}
+				count += buckets[i];
+			}
+		}
+		ASSERT(found);
+#if FASTLOG
+		return (T)(2.0 * pow(gamma, (index - offset) / fastLogger::correctingFactor) / (gamma + 1));
```

Comment:
I'm pretty sure this needs to use the cuberoot formula and not `pow`.

Reply:
I think if the cubic function is a good approximation of `log()`, then `pow` should work fine as well. I tested in various cases and `pow` works fine. That being said, happy to use the cuberoot function.